src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -319,18 +327,45 @@ public override Task WriteLineAsync(string value)
         /// <param name="disposing">True if called from Dispose; false if called from the finalizer.</param>
         protected override void Dispose(bool disposing)
         {
-            if (disposing)
+            if (!this.disposed && disposing)
             {
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
+
+                this.textWriter?.Dispose();
             }
 
-            this.textWriter.Dispose();
+            this.disposed = true;
             base.Dispose(disposing);
         }
+
+#if NETSTANDARD2_0
+        public async ValueTask DisposeAsync()
+        {
+            if (!this.disposed && this.listener != null)
+            {
+                await this.listener.StreamDisposedAsync()
+                    .ConfigureAwait(false);
+
+                this.listener = null;

```

Comment:
I think you should also call `this.textWriter.DisposeAsync()` and set it to null.

Reply:
As discussed, the BCL types in C# 7.1 do not have this method. The BCL package introduces the IAsyncDisposable interface